### PR TITLE
Multi-token trigger for event extraction implementation

### PR DESCRIPTION
**Pull request by GillesJ**
I added multi-token triggers enabling compatibility with multi-token span event trigger datasets such as ERE.

Main changes:
- Added `dygie.max_trigger_span_width` to template.jsonnet config setting: default set to 1 (single-token triggers). Behaves like `max_span_width` as trigger spans exceeding this width will be skipped by the datasetreader (same as other spans).
- dygie.py: added trigger span tuple enumeration code.
- events.py: Handle span candidates with pruning, scoring, decoding, etc. + Changed the `compute_distance_embedding` for trig_arg_embedding to work with span-to-span distance.
- document.py: trigger representations are changed to [start_tok_idx, end_tok_idx, event_label] universally so even for more single-token triggers.
- parse_ace_event.py: changed json output to trigger span format + added option for rare and mostly erroneous multitoken trigger annotations in ACE05. Default is False, so triggers are parsed as single-token, this option should probably not be used since ACE05 only has single-token triggers in principle.

Tested and working on ACE05-Event with no F1 difference with same default settings.
(If you want to test it yourself, you will have to re-parse ACE05-event from the start.)

All default usage remains the same: Documentation should not be updated except for mentioning that multitoken triggers are now supported and which config var to set. (Perhaps add warning in documentation that generating a large amount of trigger span candidates explodes memory.)

Sidenote to maintainer:
- For the trigger+argument embedding contextualization in events.py, I tested using distance_embedding (current implementation) or similarity_embeddings. Scores between distance and similarity are near-identical so using distance only is a sane default as it has a smaller embedding size. Combining distance+similarity was worse and definitely not worth the extra memory.

### CODE REVIEW

**Comment by GillesJ** [comment]
Here are the ACE05-Event scores, no significant difference with default settings in current master.
```
  "best_validation__ace-event_trig_id_precision": 0.7505518763796909,
  "best_validation__ace-event_trig_id_recall": 0.7555555555555555,
  "best_validation__ace-event_trig_id_f1": 0.753045404208195,
  "best_validation__ace-event_trig_class_precision": 0.7218543046357616,
  "best_validation__ace-event_trig_class_recall": 0.7266666666666667,
  "best_validation__ace-event_trig_class_f1": 0.7242524916943522,
  "best_validation__ace-event_arg_id_precision": 0.7365010799136069,
  "best_validation__ace-event_arg_id_recall": 0.5655058043117744,
  "best_validation__ace-event_arg_id_f1": 0.6397748592870545,
  "best_validation__ace-event_arg_class_precision": 0.67170626349892,
  "best_validation__ace-event_arg_class_recall": 0.5157545605306799,
  "best_validation__ace-event_arg_class_f1": 0.5834896810506567,
  "best_validation_loss": 310.1769244213189
```

**Comment by dwadden** [comment]
Nice, thanks! I've got a deadline next Monday, but I will review and merge after that.

File: dygie/data/dataset_readers/document.py [hunk 1]
```diff
@@ -404,6 +404,7 @@ def span_sent(self):
 
     @property
     def text(self):
+        print(self.end_sent)
```
Comment (dwadden) [inline dygie/data/dataset_readers/document.py:407]: Is this line leftover from debugging?

File: scripts/data/ace-event/parse_ace_event.py [hunk 2]
```diff
@@ -269,7 +268,7 @@ def get_token_of(doc, char):
 
 class Document:
     def __init__(self, annotation_path, text_path, doc_key, fold, heads_only=True,
-                 real_entities_only=True, include_pronouns=False, include_entity_coreference=False, include_event_coreference=False):
+                 real_entities_only=True, include_pronouns=False, include_entity_coreference=False, include_event_coreference=False, multitoken_triggers=False):
```
Comment (dwadden) [inline scripts/data/ace-event/parse_ace_event.py:271]: I make a note about this in `document.py`: It looks like you're setting `multitoken_trigger` to `False` by default, which means this code (run in default mode) won't be compatible with the updated modeling code. Is that correct? I'm fine with just setting `multitoken_trigger` to `True` by default.

**Review by dwadden — CHANGES_REQUESTED** [review]
Sorry it took me a while to do this. I took a look, everything looks generally reasonable though I haven't tested the code myself.

I make a few requests for changes in the comments below. If you don't mind, could you also report the test set results for ACE-event just like you did for validation above?

I'm trying to decide whether we should merge this into master or make it a separate branch. I think I'm fine merging, as long as you've got the bandwidth to field issues that may arise from the changes. What do you think?

File: scripts/data/ace-event/parse_ace_event.py (hunk 2 above)
Comment (dwadden) [reply scripts/data/ace-event/parse_ace_event.py:271]: Haha OK no worries. If getting your code merged in is gonna be too much of a pain, a reasonable alternative would just be for me to create a permanent `multitokentrigger` branch of `dygie` that people can use if they want to, and I can mention it in the docs. Let me know what you want to do.

**Comment by dwadden** [comment]
Just checking the status on this? @GillesJ do you want to just make your multi-token trigger stuff a separate branch?

**Comment by GillesJ** [comment]
I forgot about this. I will make a separate branch sometime this week!

**Comment by dwadden** [comment]
Sounds good!

**Comment by dwadden** [comment]
Hey @GillesJ, was doing some housekeeping and noticed this is still open. Do you think it's safe to just merge this in?

**Comment by GillesJ** [comment]
I have been away from my dev pc for a few weeks and will come back end of February. I made a mess of my local commit tree and I should do a clean merge to complete this and ensure safety. We can close this PR as I will make a new one.

**Pull request closed without merging**